### PR TITLE
UCP/TAG: Use hash table for expected tags with full mask.

### PR DESCRIPTION
**Pull request by yosefe**
- scalable matching for expected messages
- +2ns on expected tag latency
- +9ns on expected wildcard latency
- mpi_test_suite passed on 16 nodes 32ppn

### CODE REVIEW

**Comment by mellanox-github** [comment]
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/1501/ for details.

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/3425/ for details (Mellanox internal link).

**Comment by yosefe** [comment]
@brminich @evgeny-leksikov pls review

File: src/ucp/tag/tag_match.h [hunk 1]
```diff
@@ -13,6 +13,10 @@
 #include <ucs/sys/compiler_def.h>
 
 
+#define UCP_TAG_MASK_FULL           ((ucp_tag_t)-1)
+#define UCP_TAG_MATCH_HASH_SIZE     945
```
Comment (evgeny-leksikov) [inline src/ucp/tag/tag_match.h]: why 945?

File: src/ucp/tag/tag_match.h [hunk 2]
```diff
@@ -13,6 +13,10 @@
 #include <ucs/sys/compiler_def.h>
 
 
+#define UCP_TAG_MASK_FULL           ((ucp_tag_t)-1)
```
Comment (evgeny-leksikov) [inline src/ucp/tag/tag_match.h]: wouldn't it be better to use standard definition `UINT64_MAX`?

File: src/ucp/tag/tag_match.h (hunk 1 above)
Comment (yosefe) [inline src/ucp/tag/tag_match.h]: will replace with prime number, wanted to create hash table of ~1000 entries which is small enough to fit cahces

File: src/ucp/tag/tag_match.h (hunk 2 above)
Comment (yosefe) [inline src/ucp/tag/tag_match.h]: want to make sure all bits are 1-s, maybe better to write it as 0xff...ffful?

File: src/ucp/tag/tag_match.h (hunk 1 above)
Comment (evgeny-leksikov) [inline src/ucp/tag/tag_match.h]: Pls, add comment

**Comment by yosefe** [comment]
@evgeny-leksikov fixed 945 to be a prime number and added comments.
unfortunately it has small impact on latency, i think it's worth the scalability optimization.
@shamisp pls take a look as well

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/3427/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/1503/ for details.

**Comment by mellanox-github** [comment]
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/1504/ for details.

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/3428/ for details (Mellanox internal link).

File: src/ucp/tag/tag_match.c [hunk 3]
```diff
@@ -25,15 +41,75 @@ int ucp_tag_unexp_is_empty(ucp_tag_match_t *tm)
...
             return;
         }
     }
 
     ucs_bug("expected request not found");
 }
+
+static inline uint64_t ucp_tag_exp_req_seq(ucs_queue_iter_t iter)
+{
+    return (*iter == NULL) ? ULONG_MAX :
```
Comment (shamisp) [inline src/ucp/tag/tag_match.c]: how often do you hit NULL ? is there a case for unlikely ?

Comment (yosefe) [inline src/ucp/tag/tag_match.c]: NULL means we reached the end of the queue, so i don't think this is likely/unlikely case

**Comment by mellanox-github** [comment]
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/1529/ for details.

**Comment by yosefe** [comment]
@MattBBaker can you pls post ORNL failure?

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/3450/ for details (Mellanox internal link).

**Comment by yosefe** [comment]
bot:ornl:retest

**Comment by MattBBaker** [comment]
The last failure was a gcc ICE.

**Comment by shamisp** [comment]
👍

**Pull request merged**